Reduce unsafeness in InspectorThreadableLoaderClient

annevk opened the pull request:
#### 0124cf0294e7f646f860f452753a2bcbbb3c96f4
<pre>
Reduce unsafeness in InspectorThreadableLoaderClient
<a href="https://bugs.webkit.org/show_bug.cgi?id=304007">https://bugs.webkit.org/show_bug.cgi?id=304007</a>

Reviewed by Chris Dumez.

Apply <a href="https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines">https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines</a> and
address a couple tiny nits in vaguely related files.

Canonical link: <a href="https://commits.webkit.org/304665@main">https://commits.webkit.org/304665@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7e8e980cea9c7b2a63380464a13467e791ecc0d3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136240 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8596 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47520 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143950 "Built successfully") | [⏳ 🛠 win ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/59add028-df04-4846-977f-4b84b380f4c9) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/138112 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9272 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8441 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104171 "Passed tests") | [⏳ 🧪 win-tests ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/9c632ac4-bdf8-4d4f-b991-14f438929a25) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139186 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6747 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122084 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85002 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/bce9340e-be4f-4f38-878f-aba681be595a) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/6404 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4061 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4544 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115692 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40285 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146696 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8280 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40852 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/112513 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8296 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6954 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112854 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6330 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118389 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/62267 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20988 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8328 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36446 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8046 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71887 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8268 "Built successfully") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/8120 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/loader/DocumentThreadableLoader.cpp:
```diff
@@ -255,10 +257,10 @@ void DocumentThreadableLoader::cancel()
         // FIXME: This error is sent to the client in didFail(), so it should not be an internal one. Use LocalFrameLoaderClient::cancelledError() instead.
         ResourceError error(errorDomainWebKitInternal, 0, m_resource->url(), "Load cancelled"_s, ResourceError::Type::Cancellation);
-        m_client->didFail(m_document->identifier(), error); // May destroy the client.
+        client->didFail(m_document->identifier(), error); // May destroy the client.
```
This comment says `May destroy the client`. Yet, you're holding a CheckedPtr to it. Either the comment is wrong and we should remote it *or* CheckedPtr will cause crashes here and we should be the client AbstractRefCounted instead so we can ref it.

annevk replied:
Thanks! That was a lot to unravel and I'm not sure I did it all correctly. Let's see what the bots say.

annevk replied:
@cdumez given the test failures, is the right next step to pursue `class ThreadableLoaderClient : public AbstractThreadSafeRefCountedAndCanMakeWeakPtr` and also make derived classes adopt that? Or does this overall setup require a rethink by someone with more experience in the area?

cdumez replied:
I think this is going to need some work. Let me see if I can take a stab at it.

cdumez replied:
I uploaded something at https://github.com/WebKit/WebKit/pull/55386. Waiting on EWS but it's looking good locally.

annevk replied:
Thanks a lot!

I moved part of what remained of my work to https://github.com/WebKit/WebKit/pull/55557 and kept the other tiny bit of changes here.

cdumez commented on Source/WebCore/Modules/fetch/FetchResponse.cpp:
```diff
@@ -253,7 +253,7 @@ void FetchResponse::addAbortSteps(Ref<AbortSignal>&& signal)
             protectedThis->m_body->loadingFailed(*protectedThis->loadingException());
 
-        if (auto loader = WTFMove(protectedThis->m_loader))
+        if (RefPtr loader = WTFMove(protectedThis->m_loader))
```
ideally we want to use `if (RefPtr loader = std::exchange(protectedThis->m_loader, nullptr))` here. Otherwise, m_loader gets used-after-move, which works in practice but is not good practice in C++ in general.

cdumez reviewed (approved):
LGTM with nit.

annevk committed:
Reduce unsafeness in InspectorThreadableLoaderClient
https://bugs.webkit.org/show_bug.cgi?id=304007

Reviewed by Chris Dumez.

Apply https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines and
address a couple tiny nits in vaguely related files.

Canonical link: https://commits.webkit.org/304665@main
